File: src/modules/graphs/6_topologicalSorting/problems/269. Alien Dictionary.ts
/**
Input:
[
    "wrt",
    "wrf",
    "er",
    "ett",
    "rftt"
]
Output: "wertf"

Input:
[
    "z",
    "x"
]
Output: "zx"

1. You may assume all letters are in lowercase.
2. You may assume that if a is a prefix of b, then a must appear before b in the given dictionary.
3. If the order is invalid, return an empty string.
4. There may be multiple valid order of letters, return any one of them is fine.
 */

export class Solution {
    private topologicalSorting(numCourses: number, adjacencyList: number[][]) {
        const result: number[] = [];
        const inDegree: number[] = new Array(numCourses).fill(0);

        // calculating the in-degree
        for (let i = 0; i < numCourses; i++) {
            for (const neighbor of adjacencyList[i]) {
                inDegree[neighbor]++;
            }
        }

        // In the starting there will always be some nodes, with indegree of zero(always)
        const queue: number[] = [];
        for (let i = 0; i < numCourses; i++) {
            if (inDegree[i] === 0) queue.push(i);
        }

        while (queue.length > 0) {
            const vertex = queue.shift() as number;
            result.push(vertex);

            for (const neighbor of adjacencyList[vertex]) {
                if (--inDegree[neighbor] === 0) {
                    queue.push(neighbor);
                }
            }
        }

        // Check for cycles (unreachable nodes)
        if (result.length !== numCourses) {
            return []; // or return an empty string as appropriate
        }

        return result;
    }

    // k : no of vertices
    // n : no of words
    public alienOrder(words: string[], N: number, K: number): string {
        let adjacencyList = new Array(K).fill(null).map(() => []);
        for (let i = 0; i < N - 1; i++) {
            let s1 = words[i];
            let s2 = words[i + 1];
            let len = Math.min(s1.length, s2.length);
            for (let ptr = 0; ptr < len; ptr++) {
                if (s1[ptr] !== s2[ptr]) {
                    adjacencyList[s1.charCodeAt(ptr) - 97].push(s2.charCodeAt(ptr) - 97);
                    break;
                }
            }
        }
        const result = this.topologicalSorting(K, adjacencyList);
        let ans = "";
        for (const vertex of result) {
            ans += String.fromCharCode(vertex + 97);
        }
        return ans;
    }
}

File: src/modules/graphs/6_topologicalSorting/problems/269. Alien Dictionary.go
package main

import (
	"fmt"
)

type Solution struct{}

// Topological sorting function
func (sol *Solution) topologicalSorting(numCourses int, adjacencyList [][]int) []int {
	result := []int{}
	inDegree := make([]int, numCourses)

	// Calculating the in-degree
	for i := 0; i < numCourses; i++ {
		for _, neighbor := range adjacencyList[i] {
			inDegree[neighbor]++
		}
	}

	// Queue to store vertices with 0 in-degree
	queue := []int{}
	for i := 0; i < numCourses; i++ {
		if inDegree[i] == 0 {
			queue = append(queue, i)
		}
	}

	// Process vertices in the queue
	for len(queue) > 0 {
		vertex := queue[0]
		queue = queue[1:]
		result = append(result, vertex)

		for _, neighbor := range adjacencyList[vertex] {
			inDegree[neighbor]--
			if inDegree[neighbor] == 0 {
				queue = append(queue, neighbor)
			}
		}
	}

	// Check for cycles
	if len(result) != numCourses {
		return []int{} // Invalid order
	}

	return result
}

// AlienOrder function to determine the order of letters
func (sol *Solution) alienOrder(words []string, N int, K int) string {
	adjacencyList := make([][]int, K)
	for i := 0; i < K; i++ {
		adjacencyList[i] = []int{}
	}

	for i := 0; i < N-1; i++ {
		s1 := words[i]
		s2 := words[i+1]
		minLen := min(len(s1), len(s2))

		for ptr := 0; ptr < minLen; ptr++ {
			if s1[ptr] != s2[ptr] {
				adjacencyList[s1[ptr]-'a'] = append(adjacencyList[s1[ptr]-'a'], int(s2[ptr]-'a'))
				break
			}
		}
	}

	result := sol.topologicalSorting(K, adjacencyList)
	if len(result) == 0 {
		return ""
	}

	ans := ""
	for _, vertex := range result {
		ans += string(rune(vertex + 'a'))
	}

	return ans
}

// Helper function to calculate the minimum of two integers
func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	sol := &Solution{}

	// Example 1
	words1 := []string{"wrt", "wrf", "er", "ett", "rftt"}
	N1 := len(words1)
	K1 := 5
	fmt.Println(sol.alienOrder(words1, N1, K1)) // Output: "wertf"

	// Example 2
	words2 := []string{"z", "x"}
	N2 := len(words2)
	K2 := 2
	fmt.Println(sol.alienOrder(words2, N2, K2)) // Output: "zx"
}
